packages/storage/src/metadata/gunsdbMetadataStore.ts

```diff
@@ -0,0 +1,261 @@
+   */
+  private constructor(private readonly identity: Identity, gunOrServer?: GunChainReference<GunDataState> | string) {
+    if (gunOrServer) {
+      if (typeof gunOrServer === 'string') {
+        this.gun = Gun({ web: gunOrServer });
+      } else {
+        this.gun = gunOrServer;
+      }
+    } else {
+      this.gun = Gun();
+    }
+
+    this.bucketsListCache = [];
+  }
+
+  /**
+   * Creates a new instance of this metadata store for users identity.
+   *
+   * @param identity - Identity of user owning this store
+   * @param gunOrServer - initialized gun instance or peer server
+   */
+  static async fromIdentity(
+    identity: Identity,
+    gunOrServer?: GunChainReference<GunDataState> | string,
+  ): Promise<GunsdbMetadataStore> {
+    const store = new GunsdbMetadataStore(identity, gunOrServer);
+    await store.authenticateUser();
+    await store.startCachingBucketsList();
+
+    return store;
+  }
+
+  /**
+   * {@inheritDoc @spacehq/sdk#UserMetadataStore.createBucket}
+   */
+  public async createBucket(bucketSlug: string, dbId: string): Promise<BucketMetadata> {
+    // throw if dbId with bucketSlug doesn't already
+    const existingBucket = await this.findBucket(bucketSlug, dbId);
+    if (existingBucket) {
+      throw new Error('Bucket with slug and dbId already exists');
+    }
+
+    const schema: BucketMetadata = {
+      dbId,
+      encryptionKey: crypto.randomBytes(BucketEncryptionKeyLength),
+      slug: bucketSlug,
+    };
+    const encryptedMetadata = await this.encryptBucketSchema(schema);
+    const lookupKey = GunsdbMetadataStore.getBucketsLookupKey(bucketSlug, dbId);
+
+    const nodeRef = this.lookupUser.get(lookupKey).put(encryptedMetadata);
+    // store in list too. the unknown cast is required because of typescripts limitation
+    // but it to ensure that the set has a reference to the actual data
+    this.listUser.get(BucketMetadataCollection).set(nodeRef as unknown as EncryptedGunBucketMetadata);
+
+    return schema;
+  }
+
+  /**
+   * {@inheritDoc @spacehq/sdk#UserMetadataStore.findBucket}
+   */
+  public async findBucket(bucketSlug: string, dbId: string): Promise<BucketMetadata | undefined> {
+    const lookupKey = GunsdbMetadataStore.getBucketsLookupKey(bucketSlug, dbId);
+    const encryptedData = await new Promise<EncryptedGunBucketMetadata | undefined>((resolve, reject) => {
+      this.lookupUser.get(lookupKey).once((data) => {
+        resolve(data);
+      });
+    });
+    // unregister lookup
+    this.lookupUser.get(lookupKey).off();
+
+    if (!encryptedData) {
+      return undefined;
+    }
+
+    return this.decryptBucketSchema(encryptedData);
+  }
+
+  /**
+   * {@inheritDoc @spacehq/sdk#UserMetadataStore.listBuckets}
+   */
+  public async listBuckets(): Promise<BucketMetadata[]> {
+    return this.bucketsListCache;
+  }
+
+  private async startCachingBucketsList(): Promise<void> {
+    this.listUser.get(BucketMetadataCollection).map().once(async (data) => {
```

Comment:
Is there a reason to store everything in memory vs just ones that are accessed?

Reply:
This was the issue I mentioned in the PM. In order to satisfy the `listBuckets()` requirement, we need to fetch a list of all buckets created for that user, but gunsdb currently doesn't support a fetch/find all kind of query.
So this solution is to just start listening for all items in the `BucketMetadataCollection` set and aggregate them in memory, so a call to `listBuckets` will return the aggregated result.